### ISSUE TITLE
Postgres \dt

### ISSUE THREAD

**Issue by stuaxo** [xo/usql#57]
IIRC \dt and friends are implemented in the psql client ...  still, psql users will be familiar with it ... so I guess this is about adding per-database commands.

As it is I'm not sure how to list the tables, so probably go back to psql for now.

**Comment by kenshaw** [comment]
Yes, sometime before 1.0, this will be supported by `usql`. I do not have a specific ETA on that, as this is a lot of work. For all the primary databases that `usql` supports, (mysql, postgres, oracle, mssql, sqlite3), this would need to be implemented.

**Comment by dansmill** [comment]
Im very excited about this project and can see it being my main interface for sure with all the different databases that I use, but until \d family is implemented I will just keep my eye on this project.  I use them way to much.  Keep up the great work!

**Comment by kenshaw** [comment]
I have made some progress towards this. It's not done (nor will be done) until this support works for the 5 major, primary databases that `usql` targets (PostgreSQL, MySQL, Oracle, Microsoft SQL, SQLite3). Additionally, I have made the constraint that I will not release this until the logic for `\d*` is the same for `usql` as it is for `psql`. I spent a lot of time about 7 months ago writing a simple, targeted translator for PostgreSQL's `describe.c` to Go, which is available: https://github.com/xo/pgdesc but is only 80% complete at this time. There are only so many hours in the day, and limited time that I can dedicate to this. 

Hypothetically, if `usql` were the only code base I was working on, most of this could have been done well over a year ago.

**Comment by nineinchnick** [comment]
@kenshaw I'd like to give this one a shot. Do you have something in progress or can I start diving in?

**Comment by kenshaw** [comment]
@nineinchnick If this feature gets implemented, I really would want "100%" compatibility with `psql`. The way `psql` works is that the `\d` (and other informational) queries are part of the `psql` code, not the server code. As such, I started `github.com/xo/pgdesc` and had started writing some regexp based translation scripts for postgres' direct source, which I realized would not work as the logic is just too complicated to translate with regexp replacement. I really would want this generated from the postgres source at minimum, since it changes with every major postgres revision and would want to continue generating it.

I've looked into C-to-Go translation, and/or using the llvm APIs to parse the C code, but have simply not had the time to implement it in full. As for the other databases -- my assumption is that I would eventually be able to write another generator for MySQL, since its client does something similar, but the other databases would almost always need to be written by hand for compatibility. A complete implementation of this metacmd would need to support at least the 5 major databases (PostgreSQL, MySQL, MS SQL, Oracle, SQLite3), and probably should support generic `information_schema` one as well.  That would require a lot of research and manual authoring. Ideally, the major databases would all support some type of "filtering" that the `\d*` commands all support.

If you do implement this / get it done, it would be trivial to finish out the tab-completion code (for what it's worth).

Anyway, of course I'd be grateful for any well written PR that starts toward the above. Good luck.
